keyvault/client/src/main/java/com/azure/keyvault/SecretAsyncClientBuilder.java

```diff
@@ -0,0 +1,224 @@
+    private final List<HttpPipelinePolicy> policies;
+    private AsyncServiceClientCredentials credentials;
+    private HttpPipeline pipeline;
+    private URL vaultEndpoint;
+    private HttpClient httpClient;
+    private HttpLogDetailLevel httpLogDetailLevel;
+    private RetryPolicy retryPolicy;
+
+    SecretAsyncClientBuilder() {
+        retryPolicy = new RetryPolicy();
+        httpLogDetailLevel = HttpLogDetailLevel.NONE;
+        policies = new ArrayList<>();
+    }
+
+    /**
+     * Creates a {@link SecretAsyncClient} based on options set in the builder.
+     * Every time {@code build()} is called, a new instance of {@link SecretAsyncClient} is created.
+     *
+     * <p>If {@link SecretAsyncClientBuilder#pipeline(HttpPipeline) pipeline} is set, then the {@code pipeline} and
+     * {@link SecretAsyncClientBuilder#vaultEndpoint(String) serviceEndpoint} are used to create the
+     * {@link SecretAsyncClientBuilder client}. All other builder settings are ignored.</p>
+     *
+     * @return A SecretAsyncClient with the options set from the builder.
+     * @throws IllegalStateException If {@link SecretAsyncClientBuilder#credentials(AsyncServiceClientCredentials)} or
+     * {@link SecretAsyncClientBuilder#vaultEndpoint(String)} have not been set.
+     */
+    public SecretAsyncClient build() {
+
+        if (pipeline != null) {
+            return new SecretAsyncClient(vaultEndpoint, pipeline);
+        }
+
+        if (credentials == null) {
+            throw new IllegalStateException(KeyVaultErrorCodeStrings.getErrorString(KeyVaultErrorCodeStrings.CREDENTIALS_REQUIRED));
+        }
+
+        if (vaultEndpoint == null) {
+            throw new IllegalStateException(KeyVaultErrorCodeStrings.getErrorString(KeyVaultErrorCodeStrings.VAULT_END_POINT_REQUIRED));
+        }
+
+        // Closest to API goes first, closest to wire goes last.
+        final List<HttpPipelinePolicy> policies = new ArrayList<>();
+        policies.add(new UserAgentPolicy(AzureKeyVaultConfiguration.SDK_NAME, AzureKeyVaultConfiguration.SDK_VERSION));
+        policies.add(retryPolicy);
+        policies.add(new AsyncCredentialsPolicy(getAsyncTokenCredentials()));
+        policies.addAll(this.policies);
+        policies.add(new HttpLoggingPolicy(httpLogDetailLevel));
+
+        HttpPipeline pipeline = httpClient == null
+            ? new HttpPipeline(policies)
+            : new HttpPipeline(httpClient, policies);
+
+        return new SecretAsyncClient(vaultEndpoint, pipeline);
+    }
+
+    /**
+     * Sets the vault endpoint url to send HTTP requests to.
+     *
+     * @param vaultEndPoint The vault endpoint url is used as destination on Azure to send requests to.
+     * @return The updated Builder object.
+     * @throws MalformedURLException if {@code vaultEndpoint} is null or it cannot be parsed into a valid URL.
+     */
+    public SecretAsyncClientBuilder vaultEndpoint(String vaultEndPoint) throws MalformedURLException {
```

Comment:
You can't throw a checked exception here - you're forcing everyone to wrap their code. Do a try / catch and throw an IllegalArgumentException from your code instead.

Reply:
Valid point.
Updated.